python/plugins/processing/algs/qgis/FieldsMapper.py

```diff
@@ -37,150 +36,89 @@
+                                        field_def['type'],
+                                        "",
+                                        field_def['length'],
+                                        field_def['precision']))
             expression = QgsExpression(field_def['expression'])
-            expression.setGeomCalculator(da)
-            expression.setDistanceUnits(context.project().distanceUnits())
-            expression.setAreaUnits(context.project().areaUnits())
-            expression.prepare(exp_context)
             if expression.hasParserError():
-                raise GeoAlgorithmExecutionException(
+                raise QgsProcessingException(
                     self.tr(u'Parser error in expression "{}": {}')
                     .format(str(expression.expression()),
                             str(expression.parserErrorString())))
-            expressions.append(expression)
-
-        writer = output.getVectorWriter(fields, layer.wkbType(), layer.crs(), context)
-
-        # Create output vector layer with new attributes
-        error_exp = None
-        inFeat = QgsFeature()
-        outFeat = QgsFeature()
-        features = QgsProcessingUtils.getFeatures(layer, context)
-        count = QgsProcessingUtils.featureCount(layer, context)
-        if count > 0:
-            total = 100.0 / count
-            for current, inFeat in enumerate(features):
-                rownum = current + 1
-
-                geometry = inFeat.geometry()
-                outFeat.setGeometry(geometry)
-
-                attrs = []
-                for i in range(0, len(mapping)):
-                    field_def = mapping[i]
-                    expression = expressions[i]
-                    exp_context.setFeature(inFeat)
-                    exp_context.lastScope().setVariable("row_number", rownum)
-                    value = expression.evaluate(exp_context)
-                    if expression.hasEvalError():
-                        error_exp = expression
-                        break
-
-                    attrs.append(value)
-                outFeat.setAttributes(attrs)
-
-                writer.addFeature(outFeat, QgsFeatureSink.FastInsert)
-
-                feedback.setProgress(int(current * total))
-        else:
-            feedback.setProgress(100)
-
-        del writer
-
-        if error_exp is not None:
-            raise GeoAlgorithmExecutionException(
-                self.tr(u'Evaluation error in expression "{}": {}')
-                    .format(str(error_exp.expression()),
-                            str(error_exp.parserErrorString())))
+            self.expressions.append(expression)
+        return True
+
+    def outputFields(self, inputFields):
+        return self.fields
+
+    def processAlgorithm(self, parameters, context, feeback):
+        # create an expression context using thead safe processing context
+        self.expr_context = self.createExpressionContext(parameters, context)
```

Comment:
I think after this you should prepare all the expressions from self.expressions against the context - it'll dramatically speed up their evaluation.